module/icp/algs/modes/gcm.c

```diff
@@ -1036,33 +1118,73 @@ gcm_avx_will_work(void)
 }
 
 static inline void
-gcm_set_avx(boolean_t val)
+gcm_use_impl(gcm_impl impl)
 {
-	if (gcm_avx_will_work() == B_TRUE) {
-		atomic_swap_32(&gcm_use_avx, val);
+	switch (impl) {
+#if CAN_USE_GCM_ASM >= 2
+		case GCM_IMPL_AVX2:
+			if (gcm_avx2_will_work() == B_TRUE) {
+				atomic_swap_32(&gcm_impl_used, impl);
+				return;
+			}
+
+			zfs_fallthrough;
+#endif
+
+		case GCM_IMPL_AVX:
+			if (gcm_avx_will_work() == B_TRUE) {
+				atomic_swap_32(&gcm_impl_used, impl);
+				return;
+			}
+
+			zfs_fallthrough;
+
+		default:
+			atomic_swap_32(&gcm_impl_used, GCM_IMPL_GENERIC);
 	}
 }
 
 static inline boolean_t
-gcm_toggle_avx(void)
+gcm_impl_will_work(gcm_impl impl)
 {
-	if (gcm_avx_will_work() == B_TRUE) {
-		return (atomic_toggle_boolean_nv(&GCM_IMPL_USE_AVX));
-	} else {
-		return (B_FALSE);
+	switch (impl) {
+#if CAN_USE_GCM_ASM >= 2
+		case GCM_IMPL_AVX2:
+			return (gcm_avx2_will_work());
+#endif
+
+		case GCM_IMPL_AVX:
+			return (gcm_avx_will_work());
+
+		default:
+			return (B_TRUE);
 	}
 }
 
-static inline size_t
-gcm_simd_get_htab_size(boolean_t simd_mode)
+static inline gcm_impl
+gcm_toggle_impl(void)
 {
-	switch (simd_mode) {
-	case B_TRUE:
-		return (2 * 6 * 2 * sizeof (uint64_t));
+	gcm_impl current_impl, new_impl;
+	do { /* handle races */
+		current_impl = atomic_load_32(&gcm_impl_used);
+		new_impl = current_impl;
+		while (B_TRUE) { /* handle incompatble implementations */
+			new_impl = (new_impl + 1) % GCM_IMPL_MAX;
+			if (gcm_impl_will_work(new_impl)) {
+				break;
+			}
+		}
 
-	default:
-		return (0);
-	}
+	} while (atomic_cas_32(&gcm_impl_used, current_impl, new_impl) !=
+	    current_impl);
+
+	return (new_impl);
+}
+
+static inline size_t
+gcm_simd_get_htab_size(void)
+{
+	return (2 * 6 * 2 * sizeof (uint64_t));
 }
```

Comment:
This change isn't accurate. Each implementation has it's own Htab size and we have to handle it.

On lines 678 - 682 above we are allocating the Htab using this function which will return 192. But  we need 16*16=256 bytes (`const u128 Htable[16]`). The reason this works is by coincident. In the BoringSSL code I see   

> We do not use Htable[12..15].

Not sure if the removed `return 0` for the non-SIMD case is needed, will look tomorrow

Reply:
> Not sure if the removed `return 0` for the non-SIMD case is needed, will look tomorrow

Line 675-676:
```c
if (gcm_ctx->gcm_use_avx == B_TRUE) {
	size_t htab_len = gcm_simd_get_htab_size(gcm_ctx->gcm_use_avx);
```

I only ever see it called with TRUE.

I agree that relying on the fact that BoringSSL only uses the first 12 bytes is a happy coincidence; I'll look at how to rewire tomorrow.

Reply:
> > Not sure if the removed `return 0` for the non-SIMD case is needed, will look tomorrow
> I only ever see `gcm_simd_get_htab_size` called with TRUE.

Agreed.

> I agree that relying on the fact that BoringSSL only uses the first 12 bytes is a happy coincidence; I'll look at how to rewire tomorrow.

Not a big deal. You could either adjust the prototypes to use `u128[12]` or change  `gcm_simd_get_htab_size`  to return 256 for the AVX2 case. I'm fine either way. This is mostly about avoiding readers wondering about the size of Htab.